File: DevExpress-Demo/VB/NavigationDemo.Wpf/ViewModels/SolutionExplorer/SolutionExplorerModel.vb
Imports System.Collections.Generic

Namespace NavigationDemo

    Public Enum TypeNode
        File
        [Class]
        [Property]
        [Event]
        Field
        Method
        [Enum]
        EnumElement
    End Enum

    Public Class SolutionNode

        Public Property TypeNode As TypeNode

        Public Property Name As String

        Public Property TypeName As String

        Public Property FileName As String

        Private _children As List(Of SolutionNode)

        Public ReadOnly Property Children As IEnumerable(Of SolutionNode)
            Get
                Return _children
            End Get
        End Property

        Public ReadOnly Property DisplayName As String
            Get
                Return If(String.IsNullOrEmpty(TypeName), Name, String.Format("{0} : {1}", Name, TypeName))
            End Get
        End Property

        Public Property SearchString As String

        Public Property SearchName As String

        Public Sub AddChildren(ByVal child As SolutionNode)
            If _children Is Nothing Then _children = New List(Of SolutionNode)()
            _children.Add(child)
        End Sub
    End Class
End Namespace

File: DevExpress-Demo/CS/NavigationDemo.Wpf/ViewModels/SolutionExplorer/SolutionExplorerModel.cs
using System.Collections.Generic;

namespace NavigationDemo {
    public enum TypeNode {
        File,
        Class,
        Property,
        Event,
        Field,
        Method,
        Enum,
        EnumElement
    }

    public class SolutionNode {
        public TypeNode TypeNode { get; set; }

        public string Name { get; set; }

        public string TypeName { get; set; }

        public string FileName { get; set; }

        List<SolutionNode> _children;

        public IEnumerable<SolutionNode> Children { get { return _children; } }

        public string DisplayName { get { return string.IsNullOrEmpty(TypeName) ? Name : string.Format("{0} : {1}", Name, TypeName); } }

        public string SearchString { get; set; }

        public string SearchName { get; set; }


        public void AddChildren(SolutionNode child) {
            if(_children == null)
                _children = new List<SolutionNode>();
            _children.Add(child);
        }
    }
}
